File: battleship/ai.py
"""
Battleship map generator
"""
import math
import random

class ComputerPlayer:
	def __init__(self, sea_side, fleet_conf):
		self.sea_side = sea_side
		self.shots = range(self.sea_side ** 2)
		self.targets = []  # hit, but not yet sunk opponent ships

	def shot(self):
		assert len(self.shots) > 0
		for target in self.targets:
			if len(target) < 2:
				get_squares = neighbor_squares
			else:
				span = abs(target[0] - target[1])
				if span == 1:  # horisontal target
					get_squares = horizontal_neighbor_squares
				elif span == self.sea_side: # vertical target
					get_squares = vertical_neighbor_squares
				else:
					raise Exception("this is impossible!")
			for i in target:
				for shot in get_squares(i, self.sea_side):
					if shot in self.shots:
						self.shots.remove(shot)
						return (shot % 10, shot // 10)
			raise Exception("this is impossible!")
		shot = random.choice(self.shots)
		self.shots.remove(shot)
		print "removing shot", shot
		return (shot % 10, shot // 10)

	def track(self, shot, res):
		assert res in ('miss', 'hit', 'sunk')
		if res == 'miss':
			return
		index = shot[1] * self.sea_side + shot[0]
		print index, ":"
		# clear diagonal shots
		for i in diagonal_squares(index, self.sea_side):
			if i in self.shots:
				self.shots.remove(i)
				print "removing diagonal", i
		# find the target if its in not a new one
		target = None
		for t in self.targets:
			assert index not in t
			for n in neighbor_squares(index, self.sea_side):
				if n in t:
					target = t
					break
		if res == 'hit':
			if target is not None:
				target.append(index)
				return
			self.targets.append([index,])
		if res == 'sunk':
			for n in neighbor_squares(index, self.sea_side):
				if n in self.shots:
					self.shots.remove(n)
					print "removing sunk shot", n
			if target is not None:
				self.targets.remove(target)
				for i in target:
					for n in neighbor_squares(i, self.sea_side):
						if n in self.shots:
							self.shots.remove(n)
							print "removing sunk target", n

def diagonal_squares(index, side):
	res = []
	length = side ** 2
	pos = index - side + 1  # upper-right
	if pos % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	pos = index - side - 1  # upper-left
	if index % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	pos = index + side + 1  # lower-right
	if pos % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	pos = index + side - 1  # lower-left
	if index % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	return res

def neighbor_squares(index, side):
	return horizontal_neighbor_squares(index, side) + vertical_neighbor_squares(index, side)

def horizontal_neighbor_squares(index, side):
	res = []
	length = side ** 2
	pos = index - 1  # left
	if index % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	pos = index + 1  # right
	if pos % side != 0 and pos >= 0 and pos < length:
		res.append(pos)
	return res

def vertical_neighbor_squares(index, side):
	res = []
	length = side ** 2
	pos = index - side  # up
	if pos >= 0 and pos < length:
		res.append(pos)
	pos = index + side  # down
	if pos >= 0 and pos < length:
		res.append(pos)
	return res

###--------------

def setup_ships(sea_side, fleet_conf):
	"""
	typical_config = [
	(1, 4),  # one huge
	(2, 3),  # two bigs
	(3, 2),  # three mediums
	(4, 1)]  # four smalls
	"""
	ships = []
	sea = [' '] * (sea_side**2)
	for quantity, size in fleet_conf:
		for i in xrange(quantity):
			ships.append(randomly_place_ship(size, sea))
	return ships

def randomly_place_ship(size, sea):
	sea_side = int(math.sqrt(len(sea)))
	pos = generate_random_position(size, sea_side)
	while not is_squares_available(sea, pos):
		pos = generate_random_position(size, sea_side)
	occupy_squares(sea, pos)
	return pos

def generate_random_position(ship_size, sea_side):
	horizontal = random.choice((True, False))
	x = random.randrange(sea_side - (ship_size if horizontal else 0))
	y = random.randrange(sea_side - (0 if horizontal else ship_size))
	return (x, y, ship_size, horizontal)

def is_squares_available(sea, ship):
	sea_side = int(math.sqrt(len(sea)))
	x, y, ship_size, horizontal = ship
	index = y * sea_side + x
	orient = 1 if horizontal else sea_side
	for i in [index + i * orient for i in xrange(ship_size)]:
		if sea[i] != ' ':
			return False
	return True

def occupy_squares(sea, ship):
	sea_side = int(math.sqrt(len(sea)))
	x, y, ship_size, horizontal = ship
	index = y * sea_side + x
	orient = 1 if horizontal else sea_side
	pos = [index + i * orient for i in xrange(ship_size)]
	S = int(math.sqrt(len(sea)))
	for i in pos:
		assert sea[i] == ' '
		sea[i] = '#'
		# horizontal margin
		if i % S != 0 and i-1 >= 0 and i-1 < len(sea) and i-1 not in pos:
			assert sea[i-1] != '#'
			sea[i-1] = '.'  # left
		if (i+1) % S != 0 and i+1 >= 0 and i+1 < len(sea) and i+1 not in pos:
			assert sea[i+1] != '#'
			sea[i+1] = '.'  # right
		# vertical margin
		if i-S >= 0 and i-S < len(sea) and i-S not in pos:
			assert sea[i-S] != '#'
			sea[i-S] = '.'  # up
		if i+S >= 0 and i+S < len(sea) and i+S not in pos:
			assert sea[i+S] != '#'
			sea[i+S] = '.'  # down
		# diagonal margin
		if (i-S+1) % S != 0 and i-S+1 >= 0 and i-S+1 < len(sea):
			assert sea[i-S+1] != '#'
			sea[i-S+1] = '.'  # upper-right
		if i % S != 0 and i-S-1 >= 0 and i-S-1 < len(sea):
			assert sea[i-S-1] != '#'
			sea[i-S-1] = '.'  # upper-left
		if (i+S+1) % S != 0 and i+S+1 >= 0 and i+S+1 < len(sea):
			assert sea[i+S+1] != '#'
			sea[i+S+1] = '.'  # lower-right
		if i % S != 0 and i+S-1 >= 0 and i+S-1 < len(sea):
			assert sea[i+S-1] != '#'
			sea[i+S-1] = '.'  # lower-left

File: battleship/game/ai.cpp
#include "ai.h"

#include <algorithm>
#include <exception>
#include <assert.h>


namespace ai {


typedef std::vector<int> (*NEIGHBOR_FN)(size_t index, short side);
std::vector<int> neighbor_squares(size_t index, short side);
std::vector<int> horizontal_neighbor_squares(size_t index, short side);
std::vector<int> vertical_neighbor_squares(size_t index, short side);
std::vector<int> diagonal_squares(size_t index, short side);


ComputerPlayer::ComputerPlayer(short sea_side, const std::vector<logic::FleetConf>& fleet_conf)
{
	m_sea_side = sea_side;
	m_shots.resize(m_sea_side * m_sea_side);
	for (size_t i=0; i < m_shots.size(); i++) m_shots[i] = i;  // range(m_shots.size())
}


board::Pos ComputerPlayer::Shot()
{
	assert (m_shots.size() > 0);
	for (std::vector< std::vector<int> >::iterator i=m_targets.begin(); i != m_targets.end(); i++)
	{
		NEIGHBOR_FN get_squares = NULL;
		std::vector<int>& target = *i;
		if (target.size() < 2)
		{
			get_squares = neighbor_squares;
		}
		else
		{
			unsigned span = abs(target[0] - target[1]);
			if (span == 1)  // horisontal target
				get_squares = horizontal_neighbor_squares;
			else if (span == m_sea_side) // vertical target
				get_squares = vertical_neighbor_squares;
			else
				throw std::exception();  // this is impossible!
		}

		for (std::vector<int>::iterator j=target.begin(); j != target.end(); j++)
		{
			int index = *j;
			std::vector<int> squares = get_squares(index, m_sea_side);
			for (std::vector<int>::iterator n=squares.begin(); n != squares.end(); n++)
			{
				int& shot = *n;
				if (0 < std::count(m_shots.begin(), m_shots.end(), shot))
				{
					m_shots.erase(std::find(m_shots.begin(), m_shots.end(), shot));
					return board::Pos(shot % m_sea_side, shot / m_sea_side);
				}
			}
		}
		throw std::exception();  // this is impossible!
	}
	std::vector<int>::iterator shot_it = m_shots.begin() + (rand() % m_shots.size()); // random.choice(m_shots);
	int shot = *shot_it;
	m_shots.erase(shot_it);
	return board::Pos(shot % m_sea_side, shot / m_sea_side);
}




void ComputerPlayer::Track(board::Pos shot, board::SHOT res)
{
	if (res == board::MISS)
		return;
	size_t index = shot.y * m_sea_side + shot.x;
	// clear diagonal shots
	{std::vector<int> squares = diagonal_squares(index, m_sea_side);
	for (std::vector<int>::iterator i=squares.begin(); i != squares.end(); i++)
	{
		std::vector<int>::iterator find_it = std::find(m_shots.begin(), m_shots.end(), *i);
		if (find_it != m_shots.end())
			m_shots.erase(find_it);
	}}
	// find the target if its in not a new one
	std::vector< std::vector<int> >::iterator target_it = m_targets.end();
	for (std::vector< std::vector<int> >::iterator i=m_targets.begin(); i != m_targets.end(); i++)
	{
		std::vector<int>& t = *i;
		assert (0 == std::count(t.begin(), t.end(), index));
		std::vector<int> squares = neighbor_squares(index, m_sea_side);
		for (std::vector<int>::iterator square_it = squares.begin(); square_it != squares.end(); square_it++)
		{
			if (0 != std::count(t.begin(), t.end(), *square_it))
			{
				target_it = i;
				break;
			}
		}
	}

	if (res == board::HIT)
	{
		if (target_it != m_targets.end())
		{
			(*target_it).push_back(index);
			return;
		}
		m_targets.push_back(std::vector<int>(1, index));
	}

	if (res == board::SUNK)
	{
		{std::vector<int> squares = neighbor_squares(index, m_sea_side);
		for (std::vector<int>::iterator i = squares.begin(); i != squares.end(); i++)
		{
			std::vector<int>::iterator find_it = std::find(m_shots.begin(), m_shots.end(), *i);
			if (find_it != m_shots.end())
				m_shots.erase(find_it);
		}}

		if (target_it != m_targets.end())
		{
			std::vector<int> target = *target_it;
			m_targets.erase(target_it);
			for (std::vector<int>::iterator i = target.begin(); i != target.end(); i++)
			{
				std::vector<int> squares = neighbor_squares(*i, m_sea_side);
				for (std::vector<int>::iterator j=squares.begin(); j != squares.end(); j++)
				{
					std::vector<int>::iterator shot_it = std::find(m_shots.begin(), m_shots.end(), *j);
					if (shot_it != m_shots.end())
					{
						m_shots.erase(shot_it);
						return;
					}
				}
			}
		}
	}
}


std::vector<int> diagonal_squares(size_t index, short side)
{
	std::vector<int> res;
	res.reserve(4);
	unsigned length = side * side;
	int pos = index - side + 1;  // upper-right
	if (pos % side != 0 && pos >= 0 && pos < length)
	{
		res.push_back(pos);
	}
	pos = index - side - 1;  // upper-left
	if (index % side != 0 && pos >= 0 && pos < length)
	{
		res.push_back(pos);
	}
	pos = index + side + 1;  // lower-right
	if (pos % side != 0 && pos >= 0 && pos < length)
	{
		res.push_back(pos);
	}
	pos = index + side - 1;  // lower-left
	if (index % side != 0 && pos >= 0 && pos < length)
	{
		res.push_back(pos);
	}
	return res;
}

std::vector<int> neighbor_squares(size_t index, short side)
{
	std::vector<int> vec1 = horizontal_neighbor_squares(index, side);
	std::vector<int> vec2 = vertical_neighbor_squares(index, side);
	std::copy(vec2.begin(), vec2.end(), std::back_inserter(vec1));
	return vec1;
}

std::vector<int> horizontal_neighbor_squares(size_t index, short side)
{
	std::vector<int> res;
	res.reserve(4);
	size_t length = side * side;
	int pos = index - 1;  // left
	if (index % side != 0 && pos >= 0 && pos < length)
		res.push_back(pos);
	pos = index + 1;  // right
	if (pos % side != 0 && pos >= 0 && pos < length)
		res.push_back(pos);
	return res;
}

std::vector<int> vertical_neighbor_squares(size_t index, short side)
{
	std::vector<int> res;
	res.reserve(4);
	size_t length = side * side;
	int pos = index - side;  // up
	if (pos >= 0 && pos < length)
		res.push_back(pos);
	pos = index + side;  // down
	if (pos >= 0 && pos < length)
		res.push_back(pos);
	return res;
}

//--------------

board::ShipAnchor GenerateRandPos(short ship_size, short sea_side)
{
	bool horizontal = rand() % 2 == 1;  // random.choice((true, false));
	short x = rand() % (sea_side - (horizontal ? ship_size : 0));
	short y = rand() % (sea_side - (horizontal ? 0 : ship_size));
	return board::ShipAnchor(x, y, ship_size, horizontal);
}

bool IsSquaresAvailable(std::vector<char>& sea, const board::ShipAnchor& ship, unsigned sea_side)
{
	size_t index = ship.pos.y * sea_side + ship.pos.x;
	short orient = ship.horizontal ? 1 : sea_side;
	for (size_t i=0; i < ship.ship_size; i++)
	{
		if (sea[index + i * orient] != ' ')
			return false;
	}
	return true;
}

void OccupySquares(std::vector<char>& sea, const board::ShipAnchor& ship, int sea_side)
{
	int grid_len = sea.size();
	size_t start_index = ship.pos.y * sea_side + ship.pos.x;
	unsigned orient = ship.horizontal ? 1 : sea_side;

	std::vector<size_t> segments(ship.ship_size);
	for (size_t i=0; i < ship.ship_size; i++)
		segments[i] = start_index + i * orient;

	for (std::vector<size_t>::iterator i=segments.begin(); i != segments.end(); i++)
	{
		size_t& index = *i;
		assert (sea[index] == ' ');
		sea[index] = '#';
		// horizontal margin
		int pos = index - 1;  // left
		if (index % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			if (0 == std::count(segments.begin(), segments.end(), pos))
			{
				assert (sea[pos] != '#');
				sea[pos] = '.';
			}
		}
		pos = index + 1;  // right
		if (pos % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			if (0 == std::count(segments.begin(), segments.end(), pos))
			{
				assert (sea[pos] != '#');
				sea[pos] = '.';
			}
		}
		// vertical margin
		pos = index - sea_side;  // up
		if (pos >= 0 && pos < grid_len)
		{
			if (0 == std::count(segments.begin(), segments.end(), pos))
			{
				assert (sea[pos] != '#');
				sea[pos] = '.';
			}
		}
		pos = index + sea_side;  // down
		if (pos >= 0 && pos < grid_len)
		{
			if (0 == std::count(segments.begin(), segments.end(), pos))
			{
				assert (sea[pos] != '#');
				sea[pos] = '.';
			}
		}
		// diagonal margins
		pos = index - sea_side + 1;  // upper-right
		if (pos % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			assert (sea[pos] != '#');
			sea[pos] = '.';
		}
		pos = index - sea_side - 1;  // upper-left
		if (index % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			assert (sea[pos] != '#');
			sea[pos] = '.';
		}
		pos = index + sea_side + 1;  // lower-right
		if (pos % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			assert (sea[pos] != '#');
			sea[pos] = '.';
		}
		pos = index + sea_side - 1;  // lower-left
		if (index % sea_side != 0 && pos >= 0 && pos < grid_len)
		{
			assert (sea[pos] != '#');
			sea[pos] = '.';
		}
	}
}


board::ShipAnchor RandomlyPlaceShip(short size, std::vector<char>& sea, short sea_side)
{
	board::ShipAnchor pos = GenerateRandPos(size, sea_side);
	while (!IsSquaresAvailable(sea, pos, sea_side))
		pos = GenerateRandPos(size, sea_side);
	OccupySquares(sea, pos, sea_side);
	return pos;
}

std::vector<board::ShipAnchor> setup_ships(short sea_side, const std::vector<logic::FleetConf>& fleet_conf)
{
	std::vector<board::ShipAnchor> ships;
	ships.reserve(10);  // wild guess
	std::vector<char> sea(sea_side * sea_side, ' ');
	for (std::vector<logic::FleetConf>::const_iterator i=fleet_conf.begin(); i != fleet_conf.end(); i++)
	{
		for (size_t j=0; j < (*i).quantity; j++)
			ships.push_back(RandomlyPlaceShip((*i).size, sea, sea_side));
	}
	return ships;
}


}
